### PR TITLE
Reorder and merge WindowNodes across ProjectNodes

### PR DESCRIPTION
**Pull request by alandpost**
Enhance the two optimization rules that first reorder, and then merge,
adjacent WindowNodes that share the same specification to also apply
when the WindowNodes are separated by one or more ProjectNodes, when
possible.  Move the two optimizations into a single RuleSet so they can
easily share common code.  Match against a fixed number of projects in
each rule, because the pattern system doesn't yet support Kleene star.

The window function that typically produces this is LAG(X, 1), where the
lag offset constant becomes a projection below the WindowNode.

Fixes https://github.com/prestodb/presto/issues/6681

This work is further motivated by the PruneUnreferencedOutputs migration,
which introduces a new ProjectNode between nearly every pair of other
nodes.

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPullUnaryNodeAboveProjects.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+                .on(p -> {
+                    Symbol keyUsage = p.symbol("keyUsage");
+                    Symbol marker = p.symbol("marker");
+                    Symbol key = p.symbol("key");
+                    Symbol unused = p.symbol("unused");
+                    return p.project(
+                            Assignments.of(
+                                    marker, marker.toSymbolReference(),
+                                    keyUsage, p.expression("key + 1")),
+                            p.project(
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPullUnaryNodeAboveProjects.java]: Could you please indent it more, so it looks like a child to the first project?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeWindows.java [hunk 2]
```diff
@@ -199,15 +199,44 @@ public void testIdenticalWindowSpecificationsABcpA()
...
                                         project(ImmutableMap.of("ONE", expression("CAST(1 AS bigint)"), "ZERO", expression("0.0")),
-                                                window(windowMatcherBuilder -> windowMatcherBuilder
-                                                                .specification(specificationA)
-                                                                .addFunction(functionCall("sum", COMMON_FRAME, ImmutableList.of(QUANTITY_ALIAS))),
-                                                        LINEITEM_TABLESCAN_DOQSS))))));
+                                                LINEITEM_TABLESCAN_DOQSS)))));
+    }
+
+    @Test
+    public void testIdenticalWindowSpecificationsABfilterA()
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeWindows.java]: Could you please add a comment here to explain why this test is important?
It may not be obvious why this query is interesting, especially that it's connected to extra nodes which are going to be gotten rid of using `pullUnaryNodeAboveProjects`.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestReorderWindows.java [hunk 3]
```diff
@@ -215,23 +215,31 @@ public void testPrefixOfPartitionComesFirstRegardlessOfTheirOrderInSQL()
     }
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestReorderWindows.java]: Instead of removing the test for lag function, could you please keep it but just change the expected result? Similarly as you did for `TestMergeWindows`.
It'd be in order to prevent regressions (i.e. when `SwapAdjacentWindowsBySpecification` got changed breaking the `pullAboveProjections` feature).
I'd keep both tests: `lag` function test which shows that windows with lag function are reordered correctly, and one with filter to show when we don't reorder.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java [hunk 4]
```diff
@@ -41,13 +44,18 @@ public Pattern getPattern()
...
-        PlanNode child = context.getLookup().resolve(parent.getSource());
-        if (!(child instanceof WindowNode)) {
+        Optional<WindowNode> childOption = pullUnaryNodeAboveProjects(context.getLookup(), WindowNode.class, parent.getSource());
+        if (!childOption.isPresent()) {
             return Optional.empty();
         }
 
+        WindowNode child = childOption.get();
+
         if ((compare(parent, (WindowNode) child) < 0) && (!dependsOn(parent, (WindowNode) child))) {
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java]: Thanks for catching that.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPullUnaryNodeAboveProjects.java (hunk 1 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPullUnaryNodeAboveProjects.java]: This is what IntelliJ wants.  If I move the first project call `return p.project(` onto its own line, maybe you'll like the formatting better.  Let's try that!

**Comment by alandpost** [comment]
@maciejgrzybek Please take another look.  Thanks!

**Comment by alandpost** [comment]
@kokosing This one is ready for you now.  Thanks!

**Comment by kokosing** [comment]
I am a bit overwhelmed, lets postpone this after pruning rules, ok? Unless
you prefer the oposite direction.

17.07.2017 8:47 PM "Alan Post" <notifications@github.com> napisał(a):

**Comment by alandpost** [comment]
@kokosing This patch is not needed for the pruning patches I've posted so far.  It is needed for pruning WindowNode.  I have code for pruning WindowNode, but haven't posted the patch yet, because it breaks the window reorder/merge optimization.

I hope this is the information you are looking for.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 5]
```diff
@@ -115,4 +120,72 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
         }
         return Optional.of(node.replaceChildren(newChildrenBuilder.build()));
     }
+
+    /**
+     * Looks for the pattern (ProjectNode*)TargetNode, and rewrites it to TargetNode(ProjectNode*),
+     * returning an empty option if the pattern doesn't match.
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: Is it always possible to reorder such nodes? What if project nodes uses output of target node?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java [hunk 6]
```diff
@@ -44,12 +47,12 @@ public Pattern getPattern()
 
         WindowNode parent = (WindowNode) node;
 
-        PlanNode source = context.getLookup().resolve(parent.getSource());
-        if (!(source instanceof WindowNode)) {
+        Optional<WindowNode> childOption = pullUnaryNodeAboveProjects(context.getLookup(), WindowNode.class, parent.getSource());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: I think this could be extracted as separate rule, `ProjectionPushDownOverWindoNode`. We should try to avoid doing a recursion in rules.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java [hunk 7]
```diff
@@ -19,9 +19,12 @@
...
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
 
 import java.util.Optional;
 
+import static com.facebook.presto.sql.planner.iterative.rule.Util.pullUnaryNodeAboveProjects;
+import static com.facebook.presto.sql.planner.iterative.rule.Util.restrictOutputs;
 import static com.facebook.presto.sql.planner.optimizations.WindowNodeUtil.dependsOn;
 
 public class MergeAdjacentWindows
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Maybe this rule and the swap rule should belong to the same `RuleSet`? What do you think?

Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Sounds good to me -- I'm not aware of a reason that the reordering rule would be useful independently.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 5 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: I'll update the comment.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 6 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Moving the project nodes isn't a sensible standalone rule.  I'll add comments explaining why.

**Comment by alandpost** [comment]
@kokosing rebased to accommodate the new pattern stuff, with added comments to address your questions.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java [hunk 8]
```diff
@@ -46,7 +47,16 @@
     @Override
     public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
     {
-        WindowNode child = captures.get(CHILD);
+        // Pulling the descendant WindowNode above projects is done as a part of this rule, as opposed in a
+        // separate rule, because that pullup is not useful on its own, and could be undone by other rules.
+        // For example, a rule could insert a project-off node between adjacent WindowNodes that use different
+        // input symbols.
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: - have you noticed that `Lookup::resolve` got deprecated? We should try to avoid calling this. This is a last resort, to be used only if there is no other way.
 - I still do believe it should be a separate rule. Notice that we sort `WindowNode`s in the same way (`SwapAdjacentWindowsBySpecifications`), just to make them eligible to be merged. `WindowNode` sorting on its own has not much sense. This is why I suggested to put all the rules related to `WindowNode`s merging into the same `RuleSet`. To emphasis that these rules work in cooperation.
 - IMO it does not conflict with unused column pruning rules as they both push down the projection.

Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Imagine we have a rule that does the pull-up operation that the helper function does.  Then:
WA is a WindowNode that adds A = f(B).
WC is a WindowNode that adds C = f(D), with the same window specification as WA.
We start with WA over WC.
1.  Unused column pruning will introduce a ProjectNode P between them to project off D.
2.  A standalone WindowNode pull-up rule moves WC above P, adding in an identity projection for D.
3.  Any of three things are now possible:
* Elimination of P as it is now a no-op identity projection.
* Merging of WA and WC (what we wanted).
* Reinsertion of another ProjectNode between WA and WC, projecting off D again (which would block the rewrite that we wanted).

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: > 2. A standalone WindowNode pull-up rule moves WC above P, adding in an identity projection for D.

To me it does not seem to be a good rewrite, notice that P does not produce symbols which are needed by WC. They cannot be swapped. Changing assignments to me is hacky, and I am not yet sure if it is a good approach.

Anyway, If there is a conflict between window merging rules and column pruning rules then I think we need to just run them in separated iterative optimizers for a while.

We strive to support Memo with multiple nodes per group, so such alternatives that P is over WC or WC is over P will become possible in future.

Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: 1.  This patch improves the existing window-merging feature to support LAG(X, 1) which is a use case for FINRA, for example, and is thus very useful.  For a demonstration of this, see the change to TestMergeWindows.testIdenticalWindowSpecificationsAAcpA.
2.  Separating iterative optimizer rules into different optimizers defeats a main purpose of the iterative optimizer, and that is to allow the rules to work together to discover new plans that the developer might not have anticipated.
3.  I agree that simply moving WC above P is not useful, and that's why this patch does not contain a rule that would do so.  This patch only does such a move as a part of merging or reordering two window nodes.
4.  Even in the future, when the planner can explore multiple alternatives, it will be important to help steer it to explore useful alternatives, as opposed to enumerating all possible plans and hoping for the best.  There are simply too many possible plans to explore even a small fraction of them.

Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: @kokosing 
> WindowNode sorting on its own has not much sense.
It DOES make sense. The purpose of sorting is not only to make it eligible for merge but also to make windows with the same `partition by` clause adjacent, therefore we need less repartitionings to execute them, so even without `MergeAdjacentWindows`, `SwapAdjacentWindows` makes sense and speeds up some queries.
Imagine WFs as follows:
```
sum over (partition by A order by C)
avg over (partition by B)
sum over (partition by A order by D)
```
if you reorder them (`SwapAdjacentWindows` currently does it) into:
```
avg over (partition by B)
sum over (partition by A order by D)
sum over (partition by A order by C)
```
you need only 2 repartitionings instead of 3, as you'd need in the former plan.
Therefore, even though you still cannot merge any windows in such an example, the plan is better after the reordering.

I agree with @alandpost that creating a separate block of iterative rules kills benefits of iterative optimizer, preventing the rule of making window nodes mergeable in some cases which we didn't come up with while writing the rule originally.
We should keep the rules self-content and ensure they preserve the invariant of correct plan for every input the rule accepts, not more than that.
Rest should be explored by iterative-element of the optimizer.

Then it's easy to reason about the whole optimizer being correct without testing gigantic search space for all possible plans.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Ok. Thank you!

Let make it this way. First of all do not use `Lookup::resolve`. as it got deprecated. I suggest to make this rule to return a pattern like:
```
 - window
   - project
     - window
```
I believe this pattern will cover most of use cases, hoping that iterative optimizer could merge several simple projections into one.

Rules should not traverse the plan recursively. Currently it only works because `Memo` stores only a single node per group, which is going to change.

What about other nodes which could be skipped during `WindowNode` merging, like `DistintNode`, `FilterNode`?

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Also what about sorting when `WindowNode`s are separated with `ProjectNode`, like below?

```
sum over (partition by A order by C)
project()
avg over (partition by B)
project()
sum over (partition by A order by D)
```

This is why I thought that having a set of rules which will work together in order group all the `WindowNode`s together will open a bunch of opportunities for optimizer.

> Separating iterative optimizer rules into different optimizers defeats a main purpose of the iterative optimizer, and that is to allow the rules to work together to discover new plans that the developer might not have anticipated.

This is true, with a caveat it is not ;) Mostly because of `There are simply too many possible plans to explore even a small fraction of them.`. I expect to have some trade-offs like running `ReorderJoins` on its own etc. I could imagine a set of rules which could optimize window functions (it may also consists of column pruning rules, predicate pushdown, projection moving and merging rules rule and anything which could help your rules). Once `WindowNodes` are optimized (like nodes are merged or sorted) then subsequent iterative optimizers (or other optimizer stages) should not break this. `WindowNode` optimizations are not cost-based (do you expect a change here?) so the search space will not be so huge in such stage.

Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: > What about other nodes which could be skipped during WindowNode merging, like DistintNode, FilterNode?

That would be nice to handle too, however we don't see anything like that in production, so it does not seem the most important problem to solve for now.
For `FilterNode`, actually, this should be handled by other rules, i.e. pushdown.

```
Also what about sorting when WindowNodes are separated with ProjectNode, like below?

sum over (partition by A order by C)
project()
avg over (partition by B)
project()
sum over (partition by A order by D)
```

I believe what would happen then is to push down one of the projections below the WN, let's say to this intermediate stage:
```
sum over (partition by A order by C)
avg over (partition by B)
project()
project()
sum over (partition by A order by D)
```
Then, if `SwapAdjacentWindows` comes to action now it may swap first two windows but it does not matter.
Next iteration of pulling WFs above projections will pull the last WN above two projections to the form of:
```
sum over (partition by A order by C)
avg over (partition by B)
sum over (partition by A order by D)
project()
project()
```
Then merging/swapping can continue until we get:
```
avg over (partition by B)
sum over (partition by A order by D)
sum over (partition by A order by C)
project()
project()
```

Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: One minor correction to @maciejgrzybek :  FilterNode actually can't be moved, as changing the bag of rows could change the result of the window functions.  But project-on-like nodes such as SemiJoin, MarkDistinct, RowNumber, and AssignUniqueId could be moved.

An additional enhancement that could be made later would be to support moving the upper WindowNode down, as opposed to the lower WindowNode up.  This would support somewhat contrived situations.

@kokosing I don't see a reason to limit the pattern to zero-or-one project nodes between window nodes.  The algorithm would be no simpler, and the rule would be less generally effective.  Please keep in mind that the rule API exists for us to write rules -- it serves the rules, not the other way around.

**Comment by alandpost** [comment]
@kokosing now with no calls to lookup!

**Comment by kokosing** [comment]
Thank you! I will try to review it shortly.

03.09.2017 04:46 "Alan Post" <notifications@github.com> napisał(a):

@kokosing <https://github.com/kokosing> now with no calls to lookup!

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/prestodb/presto/pull/8507#issuecomment-326780868>, or mute
the thread

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java [hunk 9]
```diff
@@ -1,129 +0,0 @@
...
-import java.util.Iterator;
-import java.util.Optional;
-
-import static com.facebook.presto.matching.Capture.newCapture;
-import static com.facebook.presto.sql.planner.iterative.rule.Util.transpose;
-import static com.facebook.presto.sql.planner.optimizations.WindowNodeUtil.dependsOn;
-import static com.facebook.presto.sql.planner.plan.Patterns.source;
-import static com.facebook.presto.sql.planner.plan.Patterns.window;
-
-public class SwapAdjacentWindowsBySpecifications
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java]: It would be nice to move class in separate commit, so I could easily see what have changed in a rule.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java [hunk 10]
```diff
@@ -16,53 +16,284 @@
...
+                    .collect(toImmutableList());
+
+            return pullUnaryNodeAboveProjects(
+                    WindowNode.class,
+                    captures.get(childCapture),
+                    projects)
+                    .flatMap(newChild -> manipulateAdjacentWindowNodes(parent, newChild, context));
+        }
+
+        public abstract Optional<PlanNode> manipulateAdjacentWindowNodes(WindowNode parent, WindowNode child, Context context);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: private?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java [hunk 11]
```diff
@@ -16,53 +16,284 @@
...
+    }
+
+    public static class MergeAdjacentWindowsOverProjects
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: If you have `Swap...` working over projections do you still need to have `Merge...` working over projections as well?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java [hunk 12]
```diff
@@ -16,53 +16,284 @@
...
+                return Optional.of(
+                        restrictOutputs(context.getIdAllocator(), transposedWindows, ImmutableSet.copyOf(parent.getOutputSymbols()))
+                                .orElse(transposedWindows));
+            }
+            else {
+                return Optional.empty();
+            }
+        }
+
+        private int compare(WindowNode o1, WindowNode o2)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Please extract this as some `Comparator<WindowNode>`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java [hunk 13]
```diff
@@ -16,53 +16,284 @@
...
-
-        return Optional.of(new WindowNode(
-                parent.getId(),
-                child.getSource(),
-                parent.getSpecification(),
-                functionsBuilder.build(),
-                parent.getHashSymbol(),
-                parent.getPrePartitionedInputs(),
-                parent.getPreSortedOrderPrefix()));
+        PlanNode targetChild = target.getSources().get(0);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: use `getOnlyElement`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java [hunk 14]
```diff
@@ -16,53 +16,284 @@
...
+
+    /**
+     * Looks for the pattern (ProjectNode*)TargetNode, and rewrites it to TargetNode(ProjectNode*),
+     * returning an empty option if it can't rewrite the projects, for example because they rely on
+     * the output of the TargetNode.
+     * @param projects the nodes above the target, bottom first.
+     */
+    private static <N extends PlanNode> Optional<N> pullUnaryNodeAboveProjects(
+            Class<N> targetClass,
+            N target,
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Why not to accept `WindowNode` here?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java [hunk 15]
```diff
@@ -16,53 +16,284 @@
...
+        }
+    }
+
+    /**
+     * Looks for the pattern (ProjectNode*)TargetNode, and rewrites it to TargetNode(ProjectNode*),
+     * returning an empty option if it can't rewrite the projects, for example because they rely on
+     * the output of the TargetNode.
+     * @param projects the nodes above the target, bottom first.
+     */
+    private static <N extends PlanNode> Optional<N> pullUnaryNodeAboveProjects(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: `pullUnaryNodeAboveIdentityProjections`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 10 above)
Comment (alandpost) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: I'll make it protected, which I think is what you mean.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 12 above)
Comment (alandpost) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: How would that help?  There's nothing expecting that interface needed in the code.

In any case, this patch doesn't change any of the actual comparing or merging logic, and I think such changes could be done in a separate PR.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 11 above)
Comment (alandpost) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Yes.  WindowNodes may not need swapping.  For example, after a swap, an additional projection could get inserted between them.  Or the simple case of windows with the same frame information, but calling lag().

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 15 above)
Comment (alandpost) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: The projections aren't necessarily identity projections.  For example, in the lag() case that motivated issue #6681.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 14 above)
Comment (alandpost) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Yes, the generality is not relevant now that the function isn't in Util any more.  I'll make it specific to WindowNode, and I'll pull it in as a protected static in ManipulateAdjacentWindowsOverProjects.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 13 above)
Comment (alandpost) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: After changing to be specific to WindowNode, this code will disappear.

**Comment by alandpost** [comment]
@kokosing Please take another look.  Thanks.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java (hunk 9 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsBySpecifications.java]: What about this one?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 10 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: hm... I thought about `private`. As both belong to the same file.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 12 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Not much. I thought it would be cleaner, better separation of concepts (aka SRP),

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 15 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: so identity or constant projections only?

**Comment by kokosing** [comment]
Please address this: https://github.com/prestodb/presto/pull/8507#discussion_r136745611

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java (hunk 12 above)
Comment (maciejgrzybek) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: @kokosing It's not reusable code, so why make the impression it is? This method is meant to be used only here, it's not universal. SQL spec does not define what's the order between window nodes, so we should not make assumptions like that but suggesting WindowNodes are comparable/orderable in broader context.
The ordering is established only in order to achieve mergeability and/or lowering the repartitioning efforts, which are implementation details of Presto.
Also, before this patch we had `compare(WindowNode, WindowNode)` as a separate method, not part of class implementing `Comparator<WindowNode>`. I introduced this method recently and @martint approved it. Why this patch would require changing it?

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java]: Please, forgive me. Maybe I was not clear enough. I do not require this change. This was just my suggestion. I understand your concerns, though IMO swapping and comparing are two different things so I would separate them. 

Anyway, I will leave it up to you.

**Comment by alandpost** [comment]
@kokosing I redid the commits, putting the rules-into-ruleset change in a preliminary commit.

**Comment by kokosing** [comment]
@alandpost Thank you! Please rebase and I will merge it tomorrow (so I could respond in case of any failure).

**Comment by alandpost** [comment]
@kokosing rebased
